consensus/polybft/consensus_runtime.go

```diff
@@ -731,51 +723,37 @@ func (c *consensusRuntime) calculateUptime(currentBlock *types.Header, epoch *ep
 		return nil
 	}
 
-	firstBlockInEpoch := calculateFirstBlockOfPeriod(currentBlock.Number, c.config.PolyBFTConfig.EpochSize)
-	lastBlockInPreviousEpoch := firstBlockInEpoch - 1
-
-	startBlock := (epoch.Number * c.config.PolyBFTConfig.EpochSize) - c.config.PolyBFTConfig.EpochSize + 1
-	endBlock := getEndEpochBlockNumber(epoch.Number, c.config.PolyBFTConfig.EpochSize)
-
-	blockHeader := currentBlock
-	validators := epoch.Validators
-
-	var found bool
+	blockExtra, err := GetIbftExtra(currentBlock.ExtraData)
+	if err != nil {
+		return nil, err
+	}
 
-	for blockHeader.Number > firstBlockInEpoch {
-		if err := calculateUptimeForBlock(blockHeader, validators); err != nil {
+	// calculate uptime for current epoch
+	for blockHeader.Number > epoch.FirstBlockInEpoch {
+		if err := calculateUptimeForBlock(blockExtra, epoch.Validators); err != nil {
 			return nil, err
 		}
 
-		blockHeader, found = c.config.blockchain.GetHeaderByNumber(blockHeader.Number - 1)
-		if !found {
-			return nil, blockchain.ErrNoBlock
-		}
+		blockHeader, blockExtra, err = c.getBlockData(blockHeader.Number - 1)
 	}
 
-	// since we need to calculate uptime for the last block of the previous epoch,
-	// we need to get the validators for the that epoch from the smart contract
-	// this is something that should probably be optimized
-	if lastBlockInPreviousEpoch > 0 { // do not calculate anything for genesis block
+	// calculate uptime for blocks from previous epoch that were not processed in previous uptime
+	// since we can not calculate uptime for the last block in epoch (because of parent signatures)
+	if blockHeader.Number > uptimeLookbackSize {
 		for i := 0; i < uptimeLookbackSize; i++ {
 			validators, err := c.config.polybftBackend.GetValidators(blockHeader.Number-2, nil)
 			if err != nil {
 				return nil, err
 			}
 
-			if err := calculateUptimeForBlock(blockHeader, validators); err != nil {
+			if err := calculateUptimeForBlock(blockExtra, validators); err != nil {
 				return nil, err
 			}
 
-			blockHeader, found = c.config.blockchain.GetHeaderByNumber(blockHeader.Number - 1)
-			if !found {
-				return nil, blockchain.ErrNoBlock
-			}
+			blockHeader, blockExtra, err = c.getBlockData(blockHeader.Number - 1)
```

Comment:
This code is a bit confusing, we're overwriting `blockHeader` and `blockExtra` several times, would be great if we can simplify it.

Reply:
I'll try. Not gonna leave any promises though :sweat_smile:

Reply:
Tried it. It gets even more complicated. I'm reusing `blockHeader` and `blockExtra` because uptime is calculated like this:

**Block that is currently being built**: 20
**Parent block**: 19
**EpochSize**: 10

We start reading parent signatures from block 19 (which are the only certain way to know who signed a block), until we reach the first block in epoch, which is block 11 (This is the first for loop). This means we will not calculate uptime for the last block in epoch since we are currently building it (block is not sealed nor finalized).

To put it more graphical, uptime is calculated like this in our example (first number is block number, second number is for which block we are counting signatures when we are checking parent signatures):

======epoch3====================================================
21 (20)
================================epoch 2==========================
20 (19) 
This is where the first loop starts.
19 (18)
18 (17)
17 (16)
16 (15)
15 (14)
14 (13)
13 (12)
12 (11)
This is where the first for loop stops, and second loop begins.
11 (10)
================================epoch 1==========================
10 (9)
// This is where the second loop stops.
9 (8)

That is why we need to calculate uptime for the last two blocks of previous epoch as well. (That is the second for loop). To actually reduce the logic, I am reusing the `blockExtra` and `blockHeader` local variables, so when the first loop is done, I know from which block I need to continue (the last block of previous epoch), since their validator sets are not the same.

I can introduce new variables to make it more readable, but I think it is not necessary, and just makes the code a bit complicated and adds unnecessary lines.

What are your thoughts on this? I'll take any recommendation to improve it that's for sure. I'm just not seeing it right know :confused: .

Reply:
Ok, now I understand it better, it's still a big func so adding some other comments for improvements.

Why sorting here: https://github.com/0xPolygon/polygon-edge/pull/897/files#diff-e5b2e93010ff84c3777aef136e0834d710dd22861ca62ffbab9d5baa12a52b3cR766

Reply:
Because `uptimeCounter` is a map. We can not ensure that all validators will have the same order of addresses in the `Uptime` object lists of addresses and times they sealed a block (which is important when we calculate a hash of uptime trasnaction), so we make a slice of addresses and sort it.